### ISSUE TITLE
Incorporate Based and Power Transformer transforms for Linear Attention

### ISSUE THREAD

**Issue by andrewargatkiny** [andrewargatkiny/dense-attention#27]
We need to add promising recent linear attention transforms which are motivated by calculation of power kernel instead of exponential in attention:

1) Based (https://arxiv.org/abs/2402.18668)
2) Power Transformers / Symmetric Power Transformers (https://manifestai.com/articles/symmetric-power-transformers/ and https://arxiv.org/abs/2507.04239) / PolySketchFormer (https://arxiv.org/abs/2310.01655)

They should be implemented as new transform functions in `src/other_models/attention_kernels.py` file. These transforms should be coded in plain PyTorch.

Power Transformers, Symmetric Power Transformers and basic (exact) variant of PolySketchFormer essentially calculate the same transform. SymPow Transformer tries to bring computational costs down by combining like terms in expansion, but has a more complex algorithm and and is likely slower when implemented in pure PyTorch rather than Triton/Cuda. Therefore, we need to implement Power Transformer at first and proceed with Symmetric Power Transformers later.

The papers from 2) test the architectures only with even powers, so, while implementing general transform for arbitrary power `p`, please instantiate particular transforms in `Transform2Func` register for `p=2` and `p=4`.

**Comment by andrewargatkiny** [comment]
In the following PR, I addressed two matters of interest: 
1) Development of a working version of algorithm for odd powers;
2) Speed of initial pure PyTorch implementation.

## Odd powers

Original papers/ blogs considered even powers exclusively, due to issues with odd powers caused by small or negative values in the re-weighing mechanism of linear attention (division of attention scores by their row-wise sums).

To mitigate the issue, I devised several improvements to the Power/ BASED architectures which could be also applied to any algorithm of the linear attention class.

1) Don't use re-weighing (option `no_reweight` in config)
2) To retain numerical stability, use *post_norm* immediately after calculating attention. This can be any LayerNorm-like calculation, and it's configurable with option `no_reweight_post_norm`.
3) Additionally, use optional `scaling_d_factor`, which is calculated based on head dimension and power, to further mitigate exploding activations and gradients, when `p` or `d` are too large.

Preliminary experiments show that Power-3 Attention shows promising results on Pathfinder32 (a pretty difficult bidirectional sequence modeling task) and on causal language modeling with GPT/Llama-style architecture. Power-3 Attention is indeed better than Power-2 Attention, as proven by my experiments with Pathfinder.

## Speed-up

Initial implementation of SymPow algorithm (`SymmetricPowerEmbedding` attention kernel) was slow for training – on Pathfinder32 task it had speed around **187 seqs/s** for training and **7900 seqs/s** for inference. Params are d=128, $d_head=16$, p=2, n=1024, 6 layers). I improved on this algo, accelerating it **467 seqs/s train** and **8500 seqs/s inference**. I also developed a yet faster algorithm `SymPowFastTraining`, with **653 seqs/s train** and **8500 seqs/s inference** speed, making it best suited for training. 

However, simple `TensorPowerEmbedding` version had around **1630 seqs/s** for training and **6950-7000 seqs/s** for inference, so it's better to use this version for training, and `SymmetricPowerEmbedding` for inference. When p=3 and other params are held equal, the ratio between training speed of `SymPowFastTraining` and the simple algo gets smaller: 51 vs 91 seqs/s. For inference at p=3, difference of fastest SymPow vs Power is 1705 **seqs/s** vs 346 **seqs/s**. At some greater `p`, the SymPow algorithms should finally become the most efficient option for training, too.

I also assembled a standalone `PowerAttention` class which, under appropriate conditions, uses efficient quadratic form without `d` dim expansion. In fact, for GPT and Pathfinder training with p=3 the quadratic algo turned out to be the fastest. And when the linear form is advantageous, I set for now that it uses `TensorPowerEmbedding` as the feature map for training and `SymmetricPowerEmbedding` as the feature map for inference.
